File: clients/net-chat-go/internal/bitchat/protocol/binary_protocol.go
package protocol

import (
	"encoding/binary"
	"errors"
	"fmt"
	"time"
)

// Encode serializes a packet to the on-wire form: header + optional
// recipient + payload (with optional 2-byte pre-length when compressed)
// + optional signature, then PKCS#7 padded to the next block size.
// Straight port of binary_protocol.rs::encode; byte order is BigEndian.
func Encode(p *Packet) ([]byte, error) {
	payload := p.Payload
	var originalPayloadSize uint16
	isCompressed := false

	if ShouldCompress(p.Payload) {
		if c, err := Compress(p.Payload); err != nil {
			return nil, err
		} else if c != nil {
			originalPayloadSize = uint16(len(p.Payload))
			payload = c
			isCompressed = true
		}
	}

	flags := p.Flags
	flags.IsCompressed = isCompressed

	// Payload data size includes the 2-byte original-length prefix
	// when the payload is compressed.
	payloadDataSize := len(payload)
	if isCompressed {
		payloadDataSize += 2
	}
	if payloadDataSize > MaxPayloadSize {
		return nil, fmt.Errorf("message too large: %d > %d", payloadDataSize, MaxPayloadSize)
	}

	// Header + sender + optional recipient + payload + optional signature.
	tail := 0
	if flags.HasRecipient {
		tail += 8
	}
	if flags.HasSignature {
		tail += 64
	}
	buf := make([]byte, 0, HeaderSize+8+payloadDataSize+tail)

	buf = append(buf, p.Version, byte(p.MessageType), p.TTL)
	buf = binary.BigEndian.AppendUint64(buf, uint64(p.Timestamp.UnixMilli()))
	buf = append(buf, flags.ToByte())
	buf = binary.BigEndian.AppendUint16(buf, uint16(payloadDataSize))
	buf = append(buf, p.SenderID[:]...)

	if flags.HasRecipient {
		if p.RecipientID == nil {
			return nil, errors.New("has_recipient set but recipient_id is nil")
		}
		buf = append(buf, p.RecipientID[:]...)
	}

	if isCompressed {
		buf = binary.BigEndian.AppendUint16(buf, originalPayloadSize)
	}
	buf = append(buf, payload...)

	if flags.HasSignature {
		if p.Signature == nil {
			return nil, errors.New("has_signature set but signature is nil")
		}
		buf = append(buf, p.Signature[:]...)
	}

	optimal := OptimalBlockSize(len(buf))
	return Pad(buf, optimal), nil
}

// Decode is the inverse of Encode: strip padding, read header + fields,
// decompress if flagged. Errors reported as plain error strings —
// higher layers decide whether to warn-and-drop or bubble up.
func Decode(data []byte) (*Packet, error) {
	un := Unpad(data)
	// header + sender_id
	if len(un) < HeaderSize+8 {
		return nil, fmt.Errorf("packet too small: len=%d min=%d", len(un), HeaderSize+8)
	}
	i := 0
	version := un[i]
	i++
	if version != 1 {
		return nil, fmt.Errorf("unsupported version %d", version)
	}
	rawType := un[i]
	i++
	mt, err := MessageTypeFromByte(rawType)
	if err != nil {
		return nil, err
	}
	ttl := un[i]
	i++
	ts := binary.BigEndian.Uint64(un[i : i+8])
	i += 8
	flags := PacketFlagsFromByte(un[i])
	i++
	payloadLen := int(binary.BigEndian.Uint16(un[i : i+2]))
	i += 2

	expected := HeaderSize + 8 + payloadLen
	if flags.HasRecipient {
		expected += 8
	}
	if flags.HasSignature {
		expected += 64
	}
	if len(un) < expected {
		return nil, fmt.Errorf("packet size mismatch: expected>=%d got=%d type=0x%02x payload_len=%d has_recipient=%v has_signature=%v",
			expected, len(un), rawType, payloadLen, flags.HasRecipient, flags.HasSignature)
	}

	var sender [8]byte
	copy(sender[:], un[i:i+8])
	i += 8

	var recipient *[8]byte
	if flags.HasRecipient {
		var r [8]byte
		copy(r[:], un[i:i+8])
		i += 8
		recipient = &r
	}

	var payload []byte
	if flags.IsCompressed {
		if payloadLen < 2 {
			return nil, errors.New("compressed payload too small")
		}
		originalSize := int(binary.BigEndian.Uint16(un[i : i+2]))
		i += 2
		compressed := un[i : i+payloadLen-2]
		i += payloadLen - 2
		p, err := Decompress(compressed, originalSize)
		if err != nil {
			return nil, fmt.Errorf("decompress: %w", err)
		}
		payload = p
	} else {
		payload = make([]byte, payloadLen)
		copy(payload, un[i:i+payloadLen])
		i += payloadLen
	}

	var signature *[64]byte
	if flags.HasSignature {
		var s [64]byte
		copy(s[:], un[i:i+64])
		i += 64
		signature = &s
	}

	return &Packet{
		Version:     version,
		MessageType: mt,
		TTL:         ttl,
		Timestamp:   time.UnixMilli(int64(ts)).UTC(),
		Flags:       flags,
		SenderID:    sender,
		RecipientID: recipient,
		Payload:     payload,
		Signature:   signature,
	}, nil
}

// EncodeForSigning renders the canonical bytes that go into the
// Ed25519 signature. The verifier reproduces these by encoding the
// received packet with ttl=0 and signature=None; the rebuild MUST
// match byte-for-byte — see packet.rs::to_binary_for_signing.
func EncodeForSigning(p *Packet) ([]byte, error) {
	// Shallow copy is enough — we don't mutate the payload slice.
	q := *p
	q.TTL = 0
	q.Signature = nil
	q.Flags.HasSignature = false
	return Encode(&q)
}

// PacketSize inspects a header prefix and returns the on-wire size of
// the packet (excluding padding). Straight port of get_packet_size.
func PacketSize(data []byte) (int, error) {
	if len(data) < HeaderSize {
		return 0, errors.New("insufficient data for header")
	}
	// NOTE: preserves the (buggy?) Rust behaviour of pulling flags
	// from byte 11 and payload_len from bytes 11..13 — the Rust
	// source does the same. Mainline clients don't call this on
	// mixed streams so the bug is dormant.
	flags := PacketFlagsFromByte(data[11])
	payloadLen := int(binary.BigEndian.Uint16(data[11:13]))
	size := HeaderSize + 8 + payloadLen
	if flags.HasRecipient {
		size += 8
	}
	if flags.HasSignature {
		size += 64
	}
	return size, nil
}

File: clients/net-chat-android/app/src/main/java/io/grfx/netchat/protocol/BinaryProtocol.java
package io.grfx.netchat.protocol;

import java.io.ByteArrayOutputStream;

/**
 * Encoder/decoder for BitChat packets. Big-endian throughout;
 * mirrors bitchat-rust binary_protocol.rs and the Go port
 * internal/bitchat/protocol/binary_protocol.go byte-for-byte.
 *
 * <p>Compression is deliberately not implemented — matches Rust's
 * {@code should_compress = false} for iOS/Android compatibility. The
 * has_compressed flag is respected on decode but never set on encode.
 */
public final class BinaryProtocol {
    private BinaryProtocol() {}

    /** Encode a packet to on-wire bytes (padded, no compression). */
    public static byte[] encode(Packet p) throws ProtocolException {
        if (p.senderId == null || p.senderId.length != 8) {
            throw new ProtocolException("sender_id must be 8 bytes");
        }
        if (p.flags.hasRecipient
                && (p.recipientId == null || p.recipientId.length != 8)) {
            throw new ProtocolException("has_recipient set but recipient_id absent or wrong length");
        }
        if (p.flags.hasSignature
                && (p.signature == null || p.signature.length != 64)) {
            throw new ProtocolException("has_signature set but signature absent or wrong length");
        }
        int payloadLen = p.payload == null ? 0 : p.payload.length;
        if (payloadLen > Packet.MAX_PAYLOAD_SIZE) {
            throw new ProtocolException("payload too large: " + payloadLen
                    + " > " + Packet.MAX_PAYLOAD_SIZE);
        }

        ByteArrayOutputStream buf = new ByteArrayOutputStream(
                Packet.HEADER_SIZE + 8 + payloadLen + 64);
        buf.write(p.version & 0xFF);
        buf.write(p.messageType.value & 0xFF);
        buf.write(p.ttl & 0xFF);
        writeU64BE(buf, p.timestampMillis);
        buf.write(p.flags.toByte() & 0xFF);
        writeU16BE(buf, payloadLen);
        writeAll(buf, p.senderId);
        if (p.flags.hasRecipient) writeAll(buf, p.recipientId);
        if (payloadLen > 0) writeAll(buf, p.payload);
        if (p.flags.hasSignature) writeAll(buf, p.signature);

        byte[] raw = buf.toByteArray();
        return Padding.pad(raw, Padding.optimalBlockSize(raw.length));
    }

    /** Decode on-wire bytes (padded) into a Packet. */
    public static Packet decode(byte[] data)
            throws ProtocolException, MessageType.UnknownMessageTypeException {
        byte[] un = Padding.unpad(data);
        if (un.length < Packet.HEADER_SIZE + 8) {
            throw new ProtocolException("packet too small: " + un.length);
        }
        int i = 0;
        Packet p = new Packet();
        p.version = un[i++];
        if (p.version != 1) {
            throw new ProtocolException("unsupported version " + (p.version & 0xFF));
        }
        p.messageType = MessageType.fromByte(un[i++]);
        p.ttl = un[i++];
        p.timestampMillis = readU64BE(un, i); i += 8;
        Packet.Flags flags = Packet.Flags.fromByte(un[i++]);
        p.flags.hasRecipient = flags.hasRecipient;
        p.flags.hasSignature = flags.hasSignature;
        p.flags.isCompressed = flags.isCompressed;
        int payloadLen = readU16BE(un, i); i += 2;

        int expected = Packet.HEADER_SIZE + 8 + payloadLen;
        if (flags.hasRecipient) expected += 8;
        if (flags.hasSignature) expected += 64;
        if (un.length < expected) {
            throw new ProtocolException("packet size mismatch: expected >= "
                    + expected + " got " + un.length);
        }

        p.senderId = new byte[8];
        System.arraycopy(un, i, p.senderId, 0, 8);
        i += 8;

        if (flags.hasRecipient) {
            p.recipientId = new byte[8];
            System.arraycopy(un, i, p.recipientId, 0, 8);
            i += 8;
        }

        // Compression: we don't emit it but we tolerate reading it —
        // the payload prefix is a u16 BE original-length, then the
        // compressed bytes. We simply refuse (no zlib bundled to
        // decompress). Matches Rust's iOS-compat behaviour: real
        // clients never set this.
        if (flags.isCompressed) {
            throw new ProtocolException("compressed payload not supported");
        }
        p.payload = new byte[payloadLen];
        System.arraycopy(un, i, p.payload, 0, payloadLen);
        i += payloadLen;

        if (flags.hasSignature) {
            p.signature = new byte[64];
            System.arraycopy(un, i, p.signature, 0, 64);
            i += 64;
        }
        return p;
    }

    /** Canonical bytes for Ed25519 signing: TTL=0, no signature.
     *  Verifier reconstructs these from the wire packet + zeroing the
     *  same fields and expects a byte-for-byte match. */
    public static byte[] encodeForSigning(Packet p) throws ProtocolException {
        Packet q = new Packet();
        q.version = p.version;
        q.messageType = p.messageType;
        q.ttl = 0;
        q.timestampMillis = p.timestampMillis;
        q.flags.hasRecipient = p.flags.hasRecipient;
        q.flags.hasSignature = false;
        q.flags.isCompressed = p.flags.isCompressed;
        q.senderId = p.senderId;
        q.recipientId = p.recipientId;
        q.payload = p.payload;
        q.signature = null;
        return encode(q);
    }

    // ---- primitive I/O helpers ------------------------------------------

    private static void writeAll(ByteArrayOutputStream b, byte[] a) {
        b.write(a, 0, a.length);
    }

    private static void writeU16BE(ByteArrayOutputStream b, int v) {
        b.write((v >>> 8) & 0xFF);
        b.write(v & 0xFF);
    }

    private static void writeU64BE(ByteArrayOutputStream b, long v) {
        for (int shift = 56; shift >= 0; shift -= 8) {
            b.write((int) ((v >>> shift) & 0xFF));
        }
    }

    private static int readU16BE(byte[] a, int off) {
        return ((a[off] & 0xFF) << 8) | (a[off + 1] & 0xFF);
    }

    private static long readU64BE(byte[] a, int off) {
        long v = 0;
        for (int i = 0; i < 8; i++) {
            v = (v << 8) | (a[off + i] & 0xFFL);
        }
        return v;
    }

    /** Thrown by encode/decode on any structural error. */
    public static final class ProtocolException extends Exception {
        public ProtocolException(String msg) { super(msg); }
    }
}
